src/lib/crypto/kimchi_backend/common/dlog_plonk_based_keypair.ml

```diff
@@ -153,10 +153,21 @@ module Make (Inputs : Inputs_intf) = struct
     in
     (set_urs_info, load)
 
-  let create cs =
+  let create ~prev_challenges cs =
     let gates = Inputs.Constraint_system.finalize_and_get_gates cs in
     let public_input_size = Set_once.get_exn cs.public_input_size [%here] in
-    let index = Inputs.Index.create gates public_input_size (load_urs ()) in
+    let prev_challenges =
+      match Set_once.get cs.prev_challenges with
```

Comment:
it feels a bit weird to have a match expression on a `Set_once`. I would imagine that you always just want to `get_exn` to enforce that there's a single place that sets that value, and that it was called before. Perhaps we're trying to set `prev_challenges` a layer too high here?

Reply:
maybe `R1CS_constraint_system.create` as well as `Plonk_constraint_system.create` should take that as param?